misk/src/main/kotlin/misk/web/actions/ReadinessCheckAction.kt

```diff
@@ -37,20 +47,42 @@ class ReadinessCheckAction @Inject internal constructor(
       // Don't do healthchecks if services haven't all started. The app isn't in a good state yet,
       // and a health check could end up triggering random errors that we don't want to flood the
       // logs with.
-      return Response("", statusCode = 503)
+      return@runBlocking Response("", statusCode = 503)
+    }
+
+    // First run needs to block and get statuses
+    if (lastUpdate == null) {
+      refreshStatuses()
+    } else {
+      // This needs to be independent of the max age check so that services can recover
+      // If refreshStatuses() is never run then a failed/stale check can never recover
+      if (clock.instant().isAfter(lastUpdate!!.plusMillis(config.readiness_refresh_interval_ms.toLong()))) {
+        // Launch in the background so that we can still return the cached state
+        launch(Job()) {
+          refreshStatuses()
+        }
+      }
+
+      if (clock.instant().isAfter(lastUpdate!!.plusMillis(config.readiness_max_age_ms.toLong()))) {
```

Comment:
Would using a Guava cache here end up being simpler than rolling our own cache invalidation logic? https://github.com/google/guava/wiki/CachesExplained

Reply:
I was not aware of these, the cache described under [refresh](https://github.com/google/guava/wiki/CachesExplained#refresh) sounds exactly like what I want. Thanks for the tip!

Reply:
Hmm, @adrw I'm not sure I can get the correct semantics out of this without leaving in all of the logic that's already here. I want 4 things:

1. The initial load to fire after services have all started. This requires not running health checks on construction.
2. The cached value to be served.
3. The cache to refresh after REFRESH_TIME
4. An ERROR to happen if the cache is older than MAX_AGE

I don't see a way to get that last behavior. The `LoadingCache` will never return `null`, it will always run the `load` function. The cache does not return the age of the item. The load function cannot access the age of the item. So in order to detect Health Checks that are taking too long to run we would have to add our own stale-time tracking.